File: Day<05>/Solution.py
class Solution:
    def findK(self, a, n, m, k):
        top = 0
        left = 0
        right = m-1;
        bottom = n-1;
        dir =0;
        count=0;
        while(top<=bottom and left<=right):
            if(dir==0):
                for i in range(left, right+1):
                    k -= 1
                    if k == 0:
                        return (a[top][i])
                top += 1
                dir=1
                    
            if(dir==1):
                for i in range(top, bottom+1):
                    k -= 1
                    if k == 0:
                        return (a[i][right])
                right -= 1
                dir=2
            if(dir==2):
                for i in range(right, left - 1, -1):
                    k -= 1
                    if k == 0:
                        return (a[bottom][i])
                bottom -= 1
                dir=3
            if(dir==3):
                for i in range(bottom, top - 1, -1):
                    k -= 1
                    if k == 0:
                        return (a[i][left])
                left += 1
                dir=0
if __name__=='__main__':
    t = int(input())
    for i in range(t):
        n = list(map(int, input().strip().split()))
        arr = list(map(int, input().strip().split()))
        matrix = [[0 for i in range(n[1])]for j in range(n[0])]
        c=0
        for i in range(n[0]):
            for j in range(n[1]):
                matrix[i][j] = arr[c]
                c+=1
        obj = Solution()
        print(obj.findK(matrix, n[0], n[1], n[2]))

File: Day<05>/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution
{
  public:
	int findK(vector<vector<int>> &a, int n, int m, int k)
    {
        int top = 0;
        int left = 0;
        int right = m-1;
        int bottom = n-1;
        int dir =0;
        int count=0;
       
        while(top<=bottom and left<=right)
        {
            if(dir==0)
            {
                for(int i=left ;i<=right;i++)
                {
                    k--;
                    if(!k)
                    return (a[top][i]);
                }
                top++;
                dir=1;
                
            }
            if(dir==1)
            {
                for(int i=top;i<=bottom;i++)
                {
                    k--;
                    if(!k)
                    return (a[i][right]);
                }
                right--;
                dir=2;
                
               
            }
            if(dir==2)
            {
                for(int i=right;i>=left;i--)
                {
                    k--;
                    if(!k)
                    return (a[bottom][i]);
                }
                bottom--;
                dir=3;
            }
            if(dir==3)
            {
                for(int i=bottom;i>=top;i--)
                {
                    k--;
                    if(!k)
                    return (a[i][left]);
                }
                left++;
                dir=0;
            }
        }
    }
};

int main()
{
    int T;
    cin>>T;
  
    while(T--)
    {
        int n,m;
        int k=0;
        cin>>n>>m>>k;
        vector<vector<int>> a(n, vector<int>(m, 0));
        
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                cin>>a[i][j];
            }
        }
        Solution obj;
        cout<< obj.findK(a, n, m, k) << "\n";
    }
} 
